operator/pkg/model/ingestion/ingress.go

```diff
@@ -0,0 +1,197 @@
+	// the transform component that takes a model and outputs CiliumEnvoyConfig
+	// or other resources.
+	insecureListenerMap := make(map[string]model.HTTPListener)
+
+	sourceResource := model.FullyQualifiedResource{
+		Name:      ing.Name,
+		Namespace: ing.Namespace,
+		Group:     "",
+		Version:   "v1",
+		Kind:      "Ingress",
+	}
+
+	if ing.Spec.DefaultBackend != nil {
+		// There's a default backend set up
+
+		// get the details for the default backend
+
+		backend := model.Backend{}
+		backend.Name = ing.Spec.DefaultBackend.Service.Name
+		backend.Namespace = ing.Namespace
+
+		backend.Port = &model.BackendPort{}
+
+		if ing.Spec.DefaultBackend.Service.Port.Name != "" {
+			backend.Port.Name = ing.Spec.DefaultBackend.Service.Port.Name
+		}
+
+		if ing.Spec.DefaultBackend.Service.Port.Number != 0 {
+			backend.Port.Port = uint32(ing.Spec.DefaultBackend.Service.Port.Number)
+		}
+
+		l := model.HTTPListener{
+			Hostname: "*",
+			Routes: []model.HTTPRoute{
+				{
+					Backends: []model.Backend{
+						backend,
+					},
+				}},
+			Port: 80,
+		}
+
+		l.Sources = model.AddSource(l.Sources, sourceResource)
+
+		insecureListenerMap["*"] = l
+	}
+
+	// Now, we range across the rules, adding them in as listeners.
+	for _, rule := range ing.Spec.Rules {
+
+		host := "*"
+
+		if rule.Host != "" {
+			host = rule.Host
+		}
+
+		l, ok := insecureListenerMap[host]
+		l.Port = 80
+		l.Sources = model.AddSource(l.Sources, sourceResource)
+		if !ok {
+			l.Name = fmt.Sprintf("ing-%s-%s-%s", ing.Name, ing.Namespace, host)
+		}
+
+		l.Hostname = host
+		for _, path := range rule.HTTP.Paths {
+
+			route := model.HTTPRoute{}
+
+			switch *path.PathType {
+			case slim_networkingv1.PathTypeExact:
+				route.PathMatch.Exact = path.Path
+			case slim_networkingv1.PathTypePrefix:
+				route.PathMatch.Prefix = path.Path
+			case slim_networkingv1.PathTypeImplementationSpecific:
+				route.PathMatch.Regex = path.Path
+			default:
+				// TODO(youngnick): Figure out what to do in this case.
+				continue
+			}
+
+			backend := model.Backend{
+				Name:      path.Backend.Service.Name,
+				Namespace: ing.Namespace,
+			}
+			if path.Backend.Service != nil {
+				backend.Port = &model.BackendPort{}
+				if path.Backend.Service.Port.Name != "" {
+					backend.Port.Name = path.Backend.Service.Port.Name
+				}
+				if path.Backend.Service.Port.Number != 0 {
+					backend.Port.Port = uint32(path.Backend.Service.Port.Number)
+				}
+			}
+			route.Backends = append(route.Backends, backend)
+			l.Routes = append(l.Routes, route)
+		}
+
+		insecureListenerMap[host] = l
+	}
+
+	secureListenerMap := make(map[string]model.HTTPListener)
+
+	// First, we check for TLS config, and set them up with Listeners to return.
+	for _, tlsConfig := range ing.Spec.TLS {
+
+		for _, host := range tlsConfig.Hosts {
+
+			var l model.HTTPListener
+
+			l, ok := secureListenerMap[host]
+			if !ok {
+				l = insecureListenerMap[host]
```

Comment:
Is it possible for there to be no `insecureListenerMap[host]` entry, eg the host shows up in `ing.Spec.TLS[].Hosts[]` but not `ing.Spec.Rules[]`?

If yes, then should `l.Name` also be set in that case to match line 85? If no, then do we need `l.Sources` to be populated on line 139 since it's already populated on line 83?

Reply:
good question, the Ingress conformance tests didn't actually enforce if a particular rule is https only i.e. it's fine to have both secure and insure listeners. 

In later commit, you might see some changes related to enforceHttps, in which will return 301 http error with details about HTTPs.

Reply:
good point :eyes:. 

AFAIK, hosts in the ing.Spec.TLS need to explicitly match the host in the ing.Spec.Rules. It means that:
- the host shows up in ing.Spec.TLS[].Hosts[] but not ing.Spec.Rules[]
- the host showing up in ing.Spec.Rules[] but not in ing.Spec.TLS[]

Let me think on how to address this :thinking:

Reply:
here is the diff

<details>
<summary>diff</summary>

```diff

diff --git a/operator/pkg/model/ingestion/ingress.go b/operator/pkg/model/ingestion/ingress.go
index 8c47d1a4c4..f1edad40a9 100644
--- a/operator/pkg/model/ingestion/ingress.go
+++ b/operator/pkg/model/ingestion/ingress.go
@@ -131,7 +131,13 @@ func Ingress(ing slim_networkingv1.Ingress) []model.HTTPListener {
 
 			l, ok := secureListenerMap[host]
 			if !ok {
-				l = insecureListenerMap[host]
+				l, ok = insecureListenerMap[host]
+				if !ok {
+					l, ok = insecureListenerMap["*"]
+					if !ok {
+						continue
+					}
+				}
 			}
 
 			l.Sources = model.AddSource(l.Sources, sourceResource)
diff --git a/operator/pkg/model/ingestion/ingress_test.go b/operator/pkg/model/ingestion/ingress_test.go
index 011b7b39b0..be12dc3af5 100644
--- a/operator/pkg/model/ingestion/ingress_test.go
+++ b/operator/pkg/model/ingestion/ingress_test.go
@@ -581,7 +581,10 @@ var complexIngress = &slim_networkingv1.Ingress{
 				SecretName: "tls-very-secure-server-com",
 			},
 			{
-				Hosts:      []string{"another-very-secure.server.com"},
+				Hosts: []string{
+					"another-very-secure.server.com",
+					"not-in-use.another-very-secure.server.com",
+				},
 				SecretName: "tls-another-very-secure-server-com",
 			},
 		},
@@ -735,6 +738,64 @@ var complexIngressListeners = []model.HTTPListener{
 			},
 		},
 	},
+	{
+		Sources: []model.FullyQualifiedResource{
+			{
+				Name:      "dummy-ingress",
+				Namespace: "dummy-namespace",
+				Version:   "v1",
+				Kind:      "Ingress",
+				UID:       "d4bd3dc3-2ac5-4ab4-9dca-89c62c60177e",
+			},
+		},
+		Port:     443,
+		Hostname: "not-in-use.another-very-secure.server.com",
+		TLS: &model.TLSSecret{
+			Name:      "tls-another-very-secure-server-com",
+			Namespace: "dummy-namespace",
+		},
+		Routes: []model.HTTPRoute{
+			{
+				Backends: []model.Backend{
+					{
+						Name:      "default-backend",
+						Namespace: "dummy-namespace",
+						Port: &model.BackendPort{
+							Port: 8080,
+						},
+					},
+				},
+			},
+			{
+				PathMatch: model.StringMatch{
+					Exact: "/dummy-path",
+				},
+				Backends: []model.Backend{
+					{
+						Name:      "dummy-backend",
+						Namespace: "dummy-namespace",
+						Port: &model.BackendPort{
+							Port: 8080,
+						},
+					},
+				},
+			},
+			{
+				PathMatch: model.StringMatch{
+					Prefix: "/another-dummy-path",
+				},
+				Backends: []model.Backend{
+					{
+						Name:      "another-dummy-backend",
+						Namespace: "dummy-namespace",
+						Port: &model.BackendPort{
+							Port: 8081,
+						},
+					},
+				},
+			},
+		},
+	},
 	{
 		Sources: []model.FullyQualifiedResource{
 			{
```

</details>

Reply:
:+1: OK, so the only way to use a listener in the new TLS logic is if the earlier rules already created that listener.

This means we should be able to drop the `l.Sources = ...` statement as well, because that field was already initialized above in the rules section?

Reply:
you are right :100: